src/test/java/org/springframework/data/elasticsearch/junit/jupiter/SpringDataElasticsearchExtension.java

```diff
@@ -0,0 +1,105 @@
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.test.context.ContextConfigurationAttributes;
+import org.springframework.test.context.ContextCustomizer;
+import org.springframework.test.context.ContextCustomizerFactory;
+import org.springframework.test.context.MergedContextConfiguration;
+
+/**
+ * This extension class check in the {@link #beforeAll(ExtensionContext)} call if there is already a Elasticsearch
+ * cluster connection defined in the root store. If no, the connection to the cluster is defined according to the
+ * configuration, starting a local node if necessary. The connection is stored and will be closed when the store is
+ * shutdown at the end of all tests
+ *
+ * A ParameterResolver is implemented which enables resolving the
+ * ClusterConnectionInfo to test methods, and if a Spring context is used, the ClusterConnectionInfo can be autowired.
+ *
+ * @author Peter-Josef Meisch
+ */
+public class SpringDataElasticsearchExtension
+		implements BeforeAllCallback, ParameterResolver, ContextCustomizerFactory {
+
+	public static final String SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL = "SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL";
+	private static final Logger LOGGER = LoggerFactory.getLogger(SpringDataElasticsearchExtension.class);
+
+	private static final String STORE_KEY = ClusterConnection.class.getSimpleName();
+
+	private static final Lock initLock = new ReentrantLock();
+
+	@Override
+	public void beforeAll(ExtensionContext extensionContext) throws Exception {
+		initLock.lock();
+		try {
+			ExtensionContext rootContext = extensionContext.getRoot();
+			ExtensionContext.Store store = rootContext
+					.getStore(ExtensionContext.Namespace.create(SpringDataElasticsearchExtension.class.getName()));
+			store.getOrComputeIfAbsent(STORE_KEY, key -> {
+				LOGGER.debug("creating ClusterConnection");
+				return createClusterConnection();
+			}, ClusterConnection.class);
+		} finally {
+			initLock.unlock();
+		}
+	}
+
+	private ClusterConnection createClusterConnection() {
+		return new ClusterConnection(System.getenv(SPRING_DATA_ELASTICSEARCH_TEST_CLUSTER_URL));
+	}
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		Class<?> parameterType = parameterContext.getParameter().getType();
+		return parameterType.isAssignableFrom(ClusterConnectionInfo.class);
+	}
+
+	/*
+	 * (non javadoc)
+	 * no need to check the paramaterContext and extensionContext here, this was done before in supportsParameter.
+	 */
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		return ClusterConnection.getClusterConnectionInfo();
```

Comment:
Storing `getClusterConnectionInfo` within the `ExtensionContext.store` would remove the `ThreadLocal` dependency from the lookup.